cryptoki/src/functions/session_management.rs

```diff
@@ -28,8 +27,93 @@ impl Pkcs11 {
             ))
             .into_result()?;
         }
+        Ok(session_handle)
+    }
+
+    /// Open a new session with no callback set
+    pub fn open_session_no_callback(&self, slot_id: Slot, flags: SessionFlags) -> Result<Session> {
+        Ok(Session::new(self.open_session(slot_id, flags)?, self))
+    }
+
+    /// Opens a new management session on a token
+    ///
+    /// The purpose of a management session is to hold the state of all sessions opened on a token
+    /// by the running application.  This allows threads to open their own session, already at the
+    /// desired state set by the management session.
+    ///
+    /// # Arguments
+    ///
+    /// * `slot_id` - The slot to open the session on
+    /// * `flags` - The flags for the session
+    /// * `user_type` - The type of user to login
+    /// * `pin` - The PIN to use, or `None` if a separate authentication path exists
+    ///
+    /// # Errors
+    ///
+    /// If the session fails to open, the typical [`Error::Pkcs11`] will return.  If a management
+    /// session already exists for the token, then [`Error::ManagementSessionExists`] will be
+    /// returned.
+    pub fn open_management_session(
+        &self,
+        slot_id: Slot,
+        flags: SessionFlags,
+        user_type: UserType,
+        pin: Option<&str>,
+    ) -> Result<()> {
+        let mut map = self
+            .management_sessions
+            .lock()
+            .map_err(|_e| crate::Error::MutexPoisonError)?;
+        if let std::collections::hash_map::Entry::Vacant(_) = map.entry(slot_id) {
```

Comment:
Is it possible to replace this big `if ... else ...` by using [`contains`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.contains_key) and just returning if the key is already there?

Reply:
So the reason for this was that `clippy` threw a fit when I used `contains_key` and then after that called `insert` -> https://rust-lang.github.io/rust-clippy/master/#map_entry

Reply:
This was clippy's suggestion on how to fix the lint error

Reply:
Actually I should be using the result of Vacant() to insert something into that entry